<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
from bs4 import BeautifulSoup

In [3]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port=5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [78]:
# текст запроса
n = 10
query_3_1 = f'''select count(*) -- подсчет количества вакансий
                from public.vacancies
                limit {n}
            '''
df = pd.read_sql_query(query_3_1, connection)
df


/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/1075846225.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

In [79]:
# результат запроса
49197

49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [80]:
# текст запроса
query_3_2 = f'''select count(distinct id) --подсчет уникального количества работодателей
                from public.employers
                limit {n}
            '''
df = pd.read_sql_query(query_3_2, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/2242948043.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

In [81]:
# результат запроса
23501

23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [82]:
# текст запроса
query_3_3 = f'''select count(distinct id) --подсчет уникального количества регионов
                from public.areas
                limit {n}
            '''
df = pd.read_sql_query(query_3_3, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/918149220.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

In [83]:
# результат запроса
1362

1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [84]:
# текст запроса
query_3_5 = f'''select count (distinct id) -- подсчет количества уникальных сфер деятельности
                from public.industries
                limit {n}
            '''
df = pd.read_sql_query(query_3_5, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/2319312128.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_5, connection)


count
0    294

In [85]:
# результат запроса
294

294

***

### выводы по предварительному анализу данных
По полученным данным можно сделать следующие выводы:
1. в среднем на одного работодателя приходится 2 вакансии (это очень грубое предположение, необходим более углубленный анализ,  
так как один работодатель ищет одного работника, а какая-то фирма может набирать целый штат)
2. сферы деятельности, которых требуются работники, очень разнообразны - 294
3. так же очень богатая география регионов - 1362

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [86]:
# текст запроса
query_4_1 = f'''select count(v.name) as cnt, a.name as area  -- выводим количество вакансий применительно к каждому региону
                from public.areas a
                left join public.vacancies v on v.area_id = a.id -- присоединяем схему с вкансиями к регионам
                group by 2
                order by 1 desc
                limit {n}
            '''
df = pd.read_sql_query(query_4_1, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/1929094989.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы
5  1698     Екатеринбург
6  1670  Нижний Новгород
7  1415           Казань
8  1301        Краснодар
9  1144           Самара

In [87]:
# результат запроса
Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

NameError: name 'Москва' is not defined

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''select count(v.id)  
                from public.vacancies v
                where salary_from is not null or salary_to is not null -- прописывем условие что одно из полей не равно 0
                limit {n}
            '''
df = pd.read_sql_query(query_4_2, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_10654/1188474894.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

In [ ]:
# результат запроса
24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# текст запроса
query_4_3 = f'''select avg(salary_from) sf, avg(salary_to) st -- выводим среднее значение вилок зарплат (нижняя и верхняя) 
                from public.vacancies v
                limit {n}
            '''
df = pd.read_sql_query(query_4_3, connection)
df['sf'].round(0)
#df['st'].round(0)

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_10654/611831835.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


0    71065.0
Name: sf, dtype: float64

In [ ]:
# результат запроса
71065
110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query_4_4 = f'''select count(v.id), v.schedule, v.employment  
                from public.vacancies v
                group by 2,3
                order by 1 desc
                limit {n}
            '''
df = pd.read_sql_query(query_4_4, connection)
df.head(10)

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_10654/3273713261.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count          schedule           employment
0  35367       Полный день     Полная занятость
1   7802  Удаленная работа     Полная занятость
2   1593     Гибкий график     Полная занятость
3   1312  Удаленная работа  Частичная занятость
4    940    Сменный график     Полная занятость
5    569       Полный день           Стажировка
6    367    Вахтовый метод     Полная занятость
7    347       Полный день  Частичная занятость
8    312     Гибкий график  Частичная занятость
9    141       Полный день     Проектная работа

In [ ]:
# результат запроса
count	schedule	employment
0	35367	Полный день	Полная занятость
1	7802	Удаленная работа	Полная занятость
2	1593	Гибкий график	Полная занятость
3	1312	Удаленная работа	Частичная занятость
4	940	Сменный график	Полная занятость
5	569	Полный день	Стажировка
6	367	Вахтовый метод	Полная занятость
7	347	Полный день	Частичная занятость
8	312	Гибкий график	Частичная занятость
9	141	Полный день	Проектная работа

SyntaxError: invalid syntax (4110224063.py, line 2)

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# текст запроса
query_4_5 = f'''select count(name), experience  
                from public.vacancies
                group by 2
                order by 1
                limit {n}
            '''
df = pd.read_sql_query(query_4_5, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_10654/1597657194.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

In [ ]:
# результат запроса
count	experience
0	1337	Более 6 лет
1	7197	Нет опыта
2	14511	От 3 до 6 лет
3	26152	От 1 года до 3 лет

### выводы по детальному анализу вакансий
1. Количество вакансий прямо пропорционально величине региона (Москва, Питер и по убыванию...)
2. Ориентировочно только у половины всех вакансий есть либо нижняя либо верхняя граница заработной платы, в основном это указывает на то что никто не хочет "обещать золотые горы" до собеседования и предпочитают внести ясность в этот вопрос непосредственно на собеседовании
3. С учетом данных по нижней и верхней вилке можно сказать что средняя ЗП равна 90 тыс руб.
4. С огромным отрывом от остальных лидирует сочетание полный рабочий день - полная занятость, соотвественно и шансы трудоустроиться в таком сегменте значительно выше
5. Самый популярный показатель опыта это от 1 до 3 лет, можно сказать что работодатели предпочитают соискателй уже с минимальным опытом, но и не требуют наличия какого то огромного стажа, так-как опыт работы более 6 лет - самая малочисленная позиция 

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query_5_1 = f'''select e.name, count(v.employer_id)
                from public.EMPLOYERS e
                join public.vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
                limit 5
            '''
df = pd.read_sql_query(query_5_1, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_10654/3828994939.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

In [ ]:
# результат запроса

    name	count
0	Яндекс	      1933
1	Ростелеком	  491
2	Тинькофф	  444
3	СБЕР	      428
4	Газпром нефть 331

IndentationError: unexpected indent (2700176785.py, line 3)

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query_5_2 = f'''select a.name area, count(distinct v.id) total_vacancy, count(distinct e.id) total_employ
                from public.areas a
                left join public.employers e on a.id = e.area
                left join public.vacancies v on v.area_id = a.id
                where v.id is null
                group by a.id
                order by 3 desc
            '''
df = pd.read_sql_query(query_5_2, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12474/985139786.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area  total_vacancy  total_employ
0                Россия              0           410
1             Казахстан              0           207
2    Московская область              0            75
3    Краснодарский край              0            19
4    Ростовская область              0            18
..                  ...            ...           ...
588             Скидель              0             0
589          Белоозерск              0             0
590           Орловский              0             0
591          Павловская              0             0
592             Таксимо              0             0

[593 rows x 3 columns]

In [ ]:
# результат запроса
	area	total_vacancy	total_employ
0	Россия	0	410
1	Казахстан	0	207
2	Московская область	0	75
3	Краснодарский край	0	19
4	Ростовская область	0	18
...	...	...	...
588	Скидель	0	0
589	Белоозерск	0	0
590	Орловский	0	0
591	Павловская	0	0
592	Таксимо	0	0

IndentationError: unexpected indent (3827919587.py, line 2)

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query_5_3 = f'''select e.name empl, count(distinct v.area_id) cnt
                from public.employers e
                left join public.vacancies v on v.employer_id = e.id
                where e.name is not null
                group by 1
                order by 2 desc
                limit 5
            '''
df = pd.read_sql_query(query_5_3, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12854/3669822329.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


empl  cnt
0                  Яндекс  181
1              Ростелеком  152
2              Спецремонт  116
3  Поляков Денис Иванович   88
4               ООО ЕФИН    71

In [ ]:
# результат запроса

    empl	cnt
0	Яндекс	181
1	Ростелеком	152
2	Спецремонт	116
3	Поляков Денис Иванович	88
4	ООО ЕФИН	71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# текст запроса
query_5_4 = f'''select count(*)
                from public.employers e 
                left join public.employers_industries ei on e.id = ei.employer_id
                where ei.industry_id is null
            '''
df = pd.read_sql_query(query_5_4, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/3321336421.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

In [ ]:
# результат запроса
8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query_5_5 = f'''select e.name, count(ei.industry_id) as cnt
                from public.employers e 
                left join public.employers_industries ei on e.id = ei.employer_id
                left join public.industries i on i.id = ei.industry_id
                group by 1
                having count(ei.industry_id) = 4
                order by 1
                limit 3
            '''
df = pd.read_sql_query(query_5_5, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_12986/2217623978.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  cnt
0  101 Интернет    4
1      21vek.by    4
2          2ГИС    4

In [ ]:
# результат запроса

name	cnt
0	101 Интернет	4
1	21vek.by	4
2	2ГИС	4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [73]:
# текст запроса
query_5_6 = f'''select count (distinct e.id) id_employers --i.name name_industry
                from public.employers e 
                left join public.employers_industries ei on e.id = ei.employer_id
                left join public.industries i on i.id = ei.industry_id
                where i.name like 'Разработка программного обеспечения'
                
            '''
df = pd.read_sql_query(query_5_6, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_15080/3264687249.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


id_employers
0          3553

In [ ]:
# результат запроса
	id_employers
	3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [68]:
# код для получения списка городов-милионников
import requests
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
tables = page.find_all('tbody')
millions_table = tables[1].find_all('a')
city_list = []
for i in millions_table:
    for j in i:
        city_list.append(j)
city_list = city_list[8:]
print(city_list)        



['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [162]:
# текст запроса
query_5_7 = f'''(select a.name area, count(v.id) cnt_vacancies  
                from public.employers e
                left join public.vacancies v on e.id = v.employer_id
                left join public.areas a on a.id = v.area_id
                where e.name = 'Яндекс' and (a.name = 'Москва' or a.name = 'Санкт-Петербург' or a.name = 'Новосибирск' or a.name = 'Екатеринбург' or a.name = 'Казань' 
                or a.name = 'Нижний Новгород' or a.name = 'Красноярск' or a.name = 'Челябинск' or a.name = 'Самара' or a.name = 'Уфа' or a.name = 'Ростов-на-Дону' 
                or a.name = 'Краснодар' or a.name = 'Омск' or a.name = 'Воронеж' or a.name = 'Пермь' or a.name = 'Волгоград')
                group by 1
                order by 2 desc)
                union all
                (select 'Total', 485)    
            '''
df = pd.read_sql_query(query_5_7, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_15080/1609987941.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


area  cnt_vacancies
0            Москва             54
1   Санкт-Петербург             42
2      Екатеринбург             39
3   Нижний Новгород             36
4       Новосибирск             35
5           Воронеж             32
6         Краснодар             30
7            Самара             26
8               Уфа             26
9            Казань             25
10            Пермь             25
11   Ростов-на-Дону             25
12        Волгоград             24
13       Красноярск             23
14        Челябинск             22
15             Омск             21
16            Total            485

In [ ]:
# результат запроса
	area	cnt_vacancies
0	Москва	54
1	Санкт-Петербург	42
2	Екатеринбург	39
3	Нижний Новгород	36
4	Новосибирск	35
5	Воронеж	32
6	Краснодар	30
7	Самара	26
8	Уфа	26
9	Казань	25
10	Пермь	25
11	Ростов-на-Дону	25
12	Волгоград	24
13	Красноярск	23
14	Челябинск	22
15	Омск	21
16	Total	485

***

### Выводы по анализу работодателей
1. Как правило большое количество вакансий предлагают крупные, перспективные компании (Яндекс, Тиньков и т.д)
2. Львиная доля вакансий предлагается на территории Россиии, есть небольшое количество в Казахстане. Соотвественно количество вакансий, как правило  прямо пропорционально масштабу региона (Москва, Питер и т.д.....)
3. А что касаемо количества регионов, в которых компаниям требуются работники, то анализируя топ-5 можно сделать вывод что кроме гигантов, там располагаюся весьма неизвестные компании, это может говорить о том что компания относительно недавно сформирована и требуется большой штат с разбросом по субъектам, либо сама специфика деятельности подразумевает определенное количество работников в каждом регионе (логистические фирмы и т.д.)
4. Далеко не у всех компаний указаны сферы деятельности, что негативно влияет на дальнейший анализ данных, а также создает трудности соискателям при поиске вакансий 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [203]:
# текст запроса
query_6_1 = f'''select count(distinct v.id)  
                from public.vacancies v
                where v.name ilike '%data%' or v.name ilike '%данн%'
            '''
df = pd.read_sql_query(query_6_1, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_15080/243960555.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

In [ ]:
# результат запроса
	count
0	1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [7]:
# текст запроса
query_6_2 = f'''select count(v.name)  
                from public.vacancies v
                where(lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%' 
                or lower(v.name) like '%исследователь данных%' 
                or v.name like '%ML%' and v.name not like '%HTML%' 
                or lower(v.name) like '%machine learning%' 
                or lower(v.name) like '%машинн%обучен%')
                and (lower(v.name) like '%junior%' 
                or lower(v.experience) like 'нет опыта' 
                or lower(v.employment) like 'cтажировка')
            '''
df = pd.read_sql_query(query_6_2, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_17999/3465060275.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

In [ ]:
# результат запроса
51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [23]:
# текст запроса
query_6_3 = f'''select count(v.name)  
                from public.vacancies v
                where(lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%' 
                or lower(v.name) like '%исследователь данных%' 
                or v.name ilike '%ML%' and v.name not like '%HTML%' 
                or lower(v.name) like '%machine learning%' 
                or lower(v.name) like '%машинн%обучен%')
                and(lower(v.key_skills) like '%postgres%'
                or lower(v.key_skills) like '%sql%')
                 
                
            '''
df = pd.read_sql_query(query_6_3, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_17999/4249471036.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0    230

In [ ]:
# результат запроса
229 

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [24]:
# текст запроса
query_6_4 = f'''select count(v.name)  
                from public.vacancies v
                where(lower(v.name) like '%data scientist%' -- прописывем условия применительно к принадлежности вакансий DS
                or lower(v.name) like '%data science%' 
                or lower(v.name) like '%исследователь данных%' 
                or v.name ilike '%ML%' and v.name not like '%HTML%' 
                or lower(v.name) like '%machine learning%' 
                or lower(v.name) like '%машинн%обучен%')
                and(v.key_skills ilike '%Python%')
                 
                
            '''
df = pd.read_sql_query(query_6_4, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_17999/1783073241.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

In [ ]:
# результат запроса
	count
0	357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [9]:
# текст запроса
query_6_5 = f'''select round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)  
                from public.vacancies v
                where(lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%' 
                or lower(v.name) like '%исследователь данных%' 
                or v.name like '%ML%' and v.name not like '%HTML%' 
                or lower(v.name) like '%machine learning%' 
                or lower(v.name) like '%машинн%обучен%')
                and key_skills is not null
            '''
df = pd.read_sql_query(query_6_5, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_18175/3608577910.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

In [ ]:
# результат запроса
6,41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [21]:
# текст запроса
query_6_6 = f'''select experience exp, round(avg(coalesce(((v.salary_from + v.salary_to)/2), v.salary_from, v.salary_to)))
                from public.vacancies v
                where(lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%' 
                or lower(v.name) like '%исследователь данных%' 
                or v.name like '%ML%' and v.name not like '%HTML%' 
                or lower(v.name) like '%machine learning%' 
                or lower(v.name) like '%машинн%обучен%')
                and key_skills is not null
                and v.experience like '%3 до 6%'
                and (v.salary_from is not null or salary_to is not null)
                group by 1
            '''
df = pd.read_sql_query(query_6_6, connection)
df

/var/folders/tm/y178fb81451_vpf1w81bb4mm0000gn/T/ipykernel_19162/3898321682.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


exp     round
0  От 3 до 6 лет  243115.0

In [ ]:
# результат запроса
	exp	              round
0	От 3 до 6 лет	243115.0

In [17]:
connection.close()

### Выводы по предметному анализу
1. Итак, из общего массива вакансий (более 40 000) порядка 1700 имеют непосредственное отношение DS и аналитике данных в целом, что указывает на всотребованоость даннной специальности на рынке

2. Применительно к начинающим специалистам имеется 51 одна вакнсия, все таки компании пердпочитают сотрудников с уже имеющимся опытом
3. А вот  потребность в датасаентистах со знанием sql составляет 230 вакансий, все таки далеко не во всех случаях работодатель ставит в требованиях знание sql. Но, очевидно, здесь есть и плюсы. такие специалисты имеют больше шансов на успех (трудоустройство)

4. Требование к знанию Python указано в 357 ваконсиях, что составляет примерно четверть от общего количества. По этому факту целесообразно продублировать тезис из предыдущего вывода

5. В среднем датасаентистам необходимо иметь знания (и опыт) в 6 областях (направлениях), гле то в более узконаправленных, где в общих чертах

6. Самый интерсный вывод. Датасаентист с приличным опытом работы (от 3 до 6 лет) может рассчитывать на среднюю ЗП ри 240 000 рублей. Но тут, разумется, все опционально. Для более точного вывода необходим дополнительный анализ вакансий по регионам и работадетелям (очевидно, что Яндекс в Москве будкт плптить больше чем ООО "Рога и Копыта" в Таганроге)    

***

# Общий вывод по проекту

### Общие выводы
1. Несмотря на то что в большинстве вакансий стоит параметр минимальный опыт работы, к DS это не относится, в этой сфере работодатель предпочитает опытных специалистов

2. Средняя ЗП по всем вакансиям составила 90 000 рублей, но датасаентист с опытом может расчитывать на 200 000 +
3. География вакнсий крайне обширная, но необходимо провести доп анализ для более четкого понимания где и какие специалисты нужны. Предварительно можно сделать вывод, что чем крупнее город, тем больше вакансий в нем представлено
4. Несмотря на устоявшееся убеждение что в IT можно заниматься удаленно, по-прежнему многие компании предпочитают полный рабочий день и полную занятость 
5. Многие компании пренебрегают указывать сферы деятельности, это не очень позитивно влияет на проведения анализа данных по этому параметру

### Прогнозы и дополнительные предложения по дальнейшему исследованию
В целом неплохо было бы провести доп анализ, что бы ответить на следующие вопросы:

- Как влияет регион, где требуется работник, на уровень ЗП;

- количество вакансий предлагаемых каждой компанией (или взять хотя бы топ 10)
